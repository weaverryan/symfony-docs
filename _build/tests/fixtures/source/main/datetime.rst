.. index::
   single: Forms; Fields; DateTimeType

DateTimeType Field
==================

This field type allows the user to modify data that represents a specific
date and time (e.g. ``1984-06-05 12:15:30``).

+----------------------+-----------------------------------------------------------------------------+
| Underlying Data Type | can be ``DateTime``, string, timestamp, or array (see the ``input`` option) |
+----------------------+-----------------------------------------------------------------------------+
| Rendered as          | single text box or three select fields                                      |
+----------------------+-----------------------------------------------------------------------------+
| Options              | - `The date_format Option`_                                                 |
|                      | - `date_widget`_                                                            |
|                      | - `placeholder`_                                                            |
|                      | - `format`_                                                                 |
+----------------------+-----------------------------------------------------------------------------+
| Overridden options   | - `by_reference`_                                                           |
|                      | - `error_bubbling`_                                                         |
+----------------------+-----------------------------------------------------------------------------+
| Parent type          | :doc:`FormType </form/form_type>`                                           |
+----------------------+-----------------------------------------------------------------------------+
| Class                | :class:`Symfony\\Component\\Form\\Extension\\Core\\Type\\DateTimeType`      |
+----------------------+-----------------------------------------------------------------------------+
| Ref                  | :ref:`reference-forms-type-date-format`                                     |
|                      | :ref:`Test reference <internal-reference>`                                  |
+----------------------+-----------------------------------------------------------------------------+

Field Options
-------------

The date_format Option
~~~~~~~~~~~~~~~~~~~~~~

**type**: ``integer`` or ``string`` **default**: ``IntlDateFormatter::MEDIUM``

Defines the ``format`` option that will be passed down to the date field.
for more details.

.. tip::

    This is a little tip about something! We an also talk about specific
    methods: :method:`Symfony\\Component\\BrowserKit\\Client::doRequest`.
    Or a namespace: :namespace:`Symfony\\Component\\Validator\\Constraints`.
    Or a PHP function: :phpfunction:`parse_ini_file`.
    Or a PHP method! :phpmethod:`Locale::getDefault`.

date_widget
~~~~~~~~~~~

Date widget!

.. note::

    Sometimes we add notes. But not too often because they interrupt
    the flow.
    :ref:`internal-reference`

placeholder
~~~~~~~~~~~

.. versionadded:: 2.6
    The ``placeholder`` option was introduced in Symfony 2.6 and replaces
    ``empty_value``, which is available prior to 2.6.
    :ref:`internal-reference`

**type**: ``string`` | ``array``

If your widget option is set to ``choice``, then this field will be represented
as a series of ``select`` boxes. When the placeholder value is a string,
it will be used as the **blank value** of all select boxes::

    use Symfony\Component\Form\Extension\Core\Type\DateTimeType;

    $builder->add('startDateTime', DateTimeType::class, array(
        'placeholder' => 'Select a value',
    ));

.. seealso::

    Also check out the homepage - :doc:`/index`.
    :ref:`internal-reference`

Custom classes for links are also cool:

.. class:: list-config-options

* ``excluded_ajax_paths``
* ``intercept_redirects``
* ``position``
* ``toolbar``
* ``verbose``

format
~~~~~~

**type**: ``string`` **default**: ``Symfony\Component\Form\Extension\Core\Type\DateTimeType::HTML5_FORMAT``

If the ``widget`` option is set to ``single_text``, this option specifies
the format of the input, i.e. how Symfony will interpret the given input
as a datetime string. See `Date/Time Format Syntax`_.

.. sidebar:: Everyone loves sidebars

    But do they really? They also get in the way!

.. caution::

    Using too many sidebars or caution directives can be distracting!

time_widget
~~~~~~~~~~~

**type**: ``string`` **default**: ``choice``

Defines the ``widget`` option for the ``TimeType``.

widget
~~~~~~

**type**: ``string`` **default**: ``null``

Defines the ``widget`` option for both the ``DateType``.
and ``TimeType``. This can be overridden
with the `date_widget`_ and `time_widget`_ options.

Overridden Options
------------------

by_reference
~~~~~~~~~~~~

**default**: ``false``

The ``DateTime`` classes are treated as immutable objects.

error_bubbling
~~~~~~~~~~~~~~

**default**: ``false``

We also support code blocks!

.. code-block:: yaml

    # app/config/parameters.yml
    parameters:
        database_driver:   pdo_mysql

And configuration blocks:

.. configuration-block::

    .. code-block:: yaml

        # app/config/config.yml
        framework:
            secret:          '%secret%'
            router:          { resource: '%kernel.root_dir%/config/routing.yml' }
            # ...

        # ...

    .. code-block:: xml

        <!-- app/config/config.xml -->
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xmlns:framework="http://symfony.com/schema/dic/symfony"
            xmlns:twig="http://symfony.com/schema/dic/twig"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                http://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/symfony
                http://symfony.com/schema/dic/symfony/symfony-1.0.xsd
                http://symfony.com/schema/dic/twig
                http://symfony.com/schema/dic/twig/twig-1.0.xsd">

            <framework:config secret="%secret%">
                <framework:router resource="%kernel.root_dir%/config/routing.xml" />
                <!-- ... -->
            </framework:config>

            <!-- ... -->
        </container>

    .. code-block:: php

        // app/config/config.php
        $container->loadFromExtension('framework', array(
            'secret' => '%secret%',
            'router' => array(
                'resource' => '%kernel.root_dir%/config/routing.php',
            ),
            // ...
        ));

        // ...

Field Variables
---------------

+----------+------------+---------------------------------------+
| Variable | Type       | Usage                                 |
+==========+============+=======================================+
| widget   | ``mixed``  | The value of the ``widget`` option    |
+----------+------------+---------------------------------------+
| type     | ``string`` | Multiple lines of text here, to show  |
|          |            | that off                              |
+----------+------------+---------------------------------------+

.. _`RFC 3339`: https://tools.ietf.org/html/rfc3339
.. _`Date/Time Format Syntax`: http://userguide.icu-project.org/formatparse/datetime#TOC-Date-Time-Format-Syntax
