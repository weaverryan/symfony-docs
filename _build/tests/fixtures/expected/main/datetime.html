<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
</head>
<body>
<h1 id="datetimetype-field">DateTimeType Field<a class="headerlink" href="#datetimetype-field" title="Permalink to this headline">¶</a></h1>

<p>This field type allows the user to modify data that represents a specific
date and time (e.g. <code class="notranslate">1984-06-05 12:15:30</code>).</p>
<table class="docutils" border="1">
    <tbody valign="top">
    <tr class="row-odd">
        <td>Underlying Data Type</td>
        <td>can be <code class="notranslate">DateTime</code>, string, timestamp, or array (see the <code class="notranslate">input</code> option)</td>
    </tr>
    <tr class="row-even">
        <td>Rendered as</td>
        <td>single text box or three select fields</td>
    </tr>
    <tr class="row-odd">
        <td>Options</td>
        <td>
            <ul class="simple">
                <li><a href="#the-date-format-option" class="reference internal">The date_format Option</a></li>
                <li><a href="#date-widget" class="reference internal">date_widget</a></li>
                <li><a href="#placeholder" class="reference internal">placeholder</a></li>
                <li><a href="#format" class="reference internal">format</a></li>
            </ul>
        </td>
    </tr>
    <tr class="row-even">
        <td>Overridden options</td>
        <td>
            <ul class="simple">
                <li><a href="#by-reference" class="reference internal">by_reference</a></li>
                <li><a href="#error-bubbling" class="reference internal">error_bubbling</a></li>
            </ul>
        </td>
    </tr>
    <tr class="row-odd">
        <td>Parent type</td>
        <td><a href="form/form_type.html" class="reference internal"><em>FormType</em></a></td>
    </tr>
    <tr class="row-even">
        <td>Class</td>
        <td><a href="https://api.symfony.com/4.0/Symfony/Component/Form/Extension/Core/Type/DateTimeType.html" class="reference external" title="Symfony\Component\Form\Extension\Core\Type\DateTimeType">DateTimeType</a></td>
    </tr>
    <tr class="row-odd">
        <td>Ref</td>
        <td>
            <a href="index.html#reference-forms-type-date-format" class="reference internal"><span>Some Test Docs!</span></a>
            <a href="form/form_type.html#internal-reference" class="reference internal"><span>Test reference</span></a>
        </td>
    </tr>
    </tbody>
</table>

<h2 id="field-options">Field Options<a class="headerlink" href="#field-options" title="Permalink to this headline">¶</a></h2>

<h3 id="the-date-format-option">The date_format Option<a class="headerlink" href="#the-date-format-option" title="Permalink to this headline">¶</a></h3>

<p><strong>type</strong>: <code class="notranslate">integer</code> or <code class="notranslate">string</code> <strong>default</strong>: <code class="notranslate">IntlDateFormatter::MEDIUM</code></p>
<p>Defines the <code class="notranslate">format</code> option that will be passed down to the date field.
for more details.</p>

<div class="admonition-wrapper">
    <div class="tip"></div>
    <div class="admonition admonition-tip">
        <p class="admonition-title">Tip</p>
        <p>This is a little tip about something! We an also talk about specific
methods: <a href="https://api.symfony.com/4.0/Symfony/Component/BrowserKit/Client.html#method_doRequest" class="reference external" title="Symfony\Component\BrowserKit\Client::doRequest()">doRequest()</a>.
Or a namespace: <a href="https://api.symfony.com/4.0/Symfony/Component/Validator/Constraints.html" class="reference external" title="Symfony\Component\Validator\Constraints">Constraints</a>.
Or a PHP function: <a href="https://secure.php.net/manual/en/function.parse-ini-file.php" class="reference external" title="parse_ini_file">parse_ini_file</a>.
Or a PHP method! <a href="https://secure.php.net/manual/en/locale.getdefault.php" class="reference external" title="Locale">Locale::getDefault()</a>.</p>
    </div>
</div>

<h3 id="date-widget">date_widget<a class="headerlink" href="#date-widget" title="Permalink to this headline">¶</a></h3>
<p>Date widget!</p>
<div class="admonition-wrapper">
    <div class="note"></div>
    <div class="admonition admonition-note">
        <p class="admonition-title">Note</p>
        <p>Sometimes we add notes. But not too often because they interrupt
the flow.
<a href="form/form_type.html#internal-reference" class="reference internal"><span>FormType Documentation</span></a>
        </p>
    </div>
</div>

<h3 id="placeholder">placeholder<a class="headerlink" href="#placeholder" title="Permalink to this headline">¶</a></h3>
<div class="versionadded">
    <div><span class="versionmodified">New in version 2.6: </span><p>The <code class="notranslate">placeholder</code> option was introduced in Symfony 2.6 and replaces
<code class="notranslate">empty_value</code>, which is available prior to 2.6.
<a href="form/form_type.html#internal-reference" class="reference internal"><span>FormType Documentation</span></a></p></div>
</div>

<p><strong>type</strong>: <code class="notranslate">string</code> | <code class="notranslate">array</code></p>
<p>If your widget option is set to <code class="notranslate">choice</code>, then this field will be represented
as a series of <code class="notranslate">select</code> boxes. When the placeholder value is a string,
it will be used as the <strong>blank value</strong> of all select boxes:</p>
<div class="literal-block notranslate">
    <div class="highlight-php">
        <table class="highlighttable">
            <tr>
                <td class="linenos">
                    <div class="linenodiv"><pre> 1 2 3 4 5</pre>
                    </div>
                </td>
                <td class="code">
                    <div class="highlight">
                         <pre class="hljs php"><span class="hljs-keyword">use</span> <span class="hljs-title">Symfony</span>\<span class="hljs-title">Component</span>\<span class="hljs-title">Form</span>\<span class="hljs-title">Extension</span>\<span class="hljs-title">Core</span>\<span class="hljs-title">Type</span>\<span class="hljs-title">DateTimeType</span>; $builder-&gt;add(<span class="hljs-string">'startDateTime'</span>, DateTimeType::class, <span class="hljs-keyword">array</span>( <span class="hljs-string">'placeholder'</span> =&gt; <span class="hljs-string">'Select a value'</span>,
));</pre>
                    </div>
                </td>
            </tr>
        </table>
    </div>
</div>

<div class="admonition-wrapper"><div class="seealso"></div><div class="admonition admonition-seealso"><p>Also check out the homepage - <a href="index.html" class="reference internal"><em>Some Test Docs!</em></a>.
<a href="form/form_type.html#internal-reference" class="reference internal"><span>FormType Documentation</span></a></p>
</div></div>

<p>Custom classes for links are also cool:</p>

<ul class="simple list-config-options">
    <li><code class="notranslate">excluded_ajax_paths</code></li>
    <li><code class="notranslate">intercept_redirects</code></li>
    <li><code class="notranslate">position</code></li>
    <li><code class="notranslate">toolbar</code></li>
    <li><code class="notranslate">verbose</code></li>
</ul>

<h3 id="format">format<a class="headerlink" href="#format" title="Permalink to this headline">¶</a></h3>
<p><strong>type</strong>: <code class="notranslate">string</code> <strong>default</strong>: <code class="notranslate">Symfony\Component\Form\Extension\Core\Type\DateTimeType::HTML5_FORMAT</code></p>
<p>If the <code class="notranslate">widget</code> option is set to <code class="notranslate">single_text</code>, this option specifies
the format of the input, i.e. how Symfony will interpret the given input
as a datetime string. See <a href="http://userguide.icu-project.org/formatparse/datetime#TOC-Date-Time-Format-Syntax" class="reference external">Date/Time Format Syntax</a>.</p>

<div class="admonition-wrapper">
    <div class="sidebar"></div>
    <div class="admonition admonition-sidebar">
        <p class="sidebar-title">Everyone loves sidebars</p>
        <p>But do they really? They also get in the way!</p>
    </div>
</div>


<div class="admonition-wrapper">
    <div class="caution"></div>
    <div class="admonition admonition-caution">
        <p class="admonition-title">Caution</p>
        <p>Using too many sidebars or caution directives can be distracting!</p>
    </div>
</div>

<h3 id="time-widget">time_widget<a class="headerlink" href="#time-widget" title="Permalink to this headline">¶</a></h3>

<p><strong>type</strong>: <code class="notranslate">string</code> <strong>default</strong>: <code class="notranslate">choice</code></p>
<p>Defines the <code class="notranslate">widget</code> option for the <code class="notranslate">TimeType</code>.</p>

<h3 id="widget">widget<a class="headerlink" href="#widget" title="Permalink to this headline">¶</a></h3>

<p><strong>type</strong>: <code class="notranslate">string</code> <strong>default</strong>: <code class="notranslate">null</code></p>
<p>Defines the <code class="notranslate">widget</code> option for both the <code class="notranslate">DateType</code>.
and <code class="notranslate">TimeType</code>. This can be overridden
with the <a href="#date-widget" class="reference internal">date_widget</a> and <a href="#time-widget" class="reference internal">time_widget</a> options.</p>

<h2 id="overridden-options">Overridden Options<a class="headerlink" href="#overridden-options" title="Permalink to this headline">¶</a></h2>

<h3 id="by-reference">by_reference<a class="headerlink" href="#by-reference" title="Permalink to this headline">¶</a></h3>

<p><strong>default</strong>: <code class="notranslate">false</code></p>
<p>The <code class="notranslate">DateTime</code> classes are treated as immutable objects.</p>

<h3 id="error-bubbling">error_bubbling<a class="headerlink" href="#error-bubbling" title="Permalink to this headline">¶</a></h3>
<p><strong>default</strong>: <code class="notranslate">false</code></p>
<p>We also support code blocks!</p>
<div class="literal-block notranslate">
    <div class="highlight-yaml">
        <table class="highlighttable">
            <tr>
                <td class="linenos">
                    <div class="linenodiv"><pre> 1 2 3</pre>
                    </div>
                </td>
                <td class="code">
                    <div class="highlight">
                        <pre class="hljs yaml"><span class="hljs-comment"># app/config/parameters.yml</span>
<span class="hljs-attr">parameters:</span>
<span class="hljs-attr"> database_driver:</span> <span class="hljs-string">pdo_mysql</span></pre>
                    </div>
                </td>
            </tr>
        </table>
    </div>
</div>

<p>And configuration blocks:</p>
<div class="configuration-block">
    <ul class="simple">
        <li><em>YAML</em>
            <div class="literal-block notranslate">
                <div class="highlight-yaml">
                    <table class="highlighttable">
                        <tr>
                            <td class="linenos">
                                <div class="linenodiv"><pre> 1 2 3 4 5 6 7</pre>
                                </div>
                            </td>
                            <td class="code">
                                <div class="highlight">
                                    <pre class="hljs yaml"><span class="hljs-comment"># app/config/config.yml</span>
<span class="hljs-attr">framework:</span>
<span class="hljs-attr"> secret:</span> <span class="hljs-string">'%secret%'</span>
<span class="hljs-attr"> router:</span> <span class="hljs-string">{</span> <span class="hljs-attr">resource:</span> <span class="hljs-string">'%kernel.root_dir%/config/routing.yml'</span> <span class="hljs-string">}</span> <span class="hljs-comment"># ...</span> <span class="hljs-comment"># ...</span></pre>
                                </div>
                            </td>
                        </tr>
                    </table>
                </div>
            </div>
        </li>
        <li><em>XML</em>
            <div class="literal-block notranslate">
                <div class="highlight-xml">
                    <table class="highlighttable">
                        <tr>
                            <td class="linenos">
                                <div class="linenodiv"><pre> 1 2 3 4 5 6 7 8 9
10
11
12
13
14
15
16
17
18
19
20</pre>
                                </div>
                            </td>
                            <td class="code">
                                <div class="highlight">
                                    <pre class="hljs xml"><span class="hljs-comment">&lt;!-- app/config/config.xml --&gt;</span>
<span class="php"><span class="hljs-meta">&lt;?</span>xml version=<span class="hljs-string">"1.0"</span> encoding=<span class="hljs-string">"UTF-8"</span> <span class="hljs-meta">?&gt;</span></span>
<span class="hljs-tag">&lt;<span class="hljs-name">container</span> <span class="hljs-attr">xmlns</span>=<span class="hljs-string">"http://symfony.com/schema/dic/services"</span>
    <span class="hljs-attr">xmlns:xsi</span>=<span class="hljs-string">"http://www.w3.org/2001/XMLSchema-instance"</span>
    <span class="hljs-attr">xmlns:framework</span>=<span class="hljs-string">"http://symfony.com/schema/dic/symfony"</span>
    <span class="hljs-attr">xmlns:twig</span>=<span class="hljs-string">"http://symfony.com/schema/dic/twig"</span>
    <span class="hljs-attr">xsi:schemaLocation</span>=<span class="hljs-string">"http://symfony.com/schema/dic/services
        http://symfony.com/schema/dic/services/services-1.0.xsd
        http://symfony.com/schema/dic/symfony
        http://symfony.com/schema/dic/symfony/symfony-1.0.xsd
        http://symfony.com/schema/dic/twig
        http://symfony.com/schema/dic/twig/twig-1.0.xsd"</span>&gt;</span>

    <span class="hljs-tag">&lt;<span class="hljs-name">framework:config</span> <span class="hljs-attr">secret</span>=<span class="hljs-string">"%secret%"</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-name">framework:router</span> <span class="hljs-attr">resource</span>=<span class="hljs-string">"%kernel.root_dir%/config/routing.xml"</span> /&gt;</span>
        <span class="hljs-comment">&lt;!-- ... --&gt;</span>
    <span class="hljs-tag">&lt;/<span class="hljs-name">framework:config</span>&gt;</span>

    <span class="hljs-comment">&lt;!-- ... --&gt;</span>
<span class="hljs-tag">&lt;/<span class="hljs-name">container</span>&gt;</span></pre>
                                </div>
                            </td>
                        </tr>
                    </table>
                </div>
            </div>
        </li>
        <li><em>PHP</em>
            <div class="literal-block notranslate">
                <div class="highlight-php">
                    <table class="highlighttable">
                        <tr>
                            <td class="linenos">
                                <div class="linenodiv"><pre> 1 2 3 4 5 6 7 8 9
10</pre>
                                </div>
                            </td>
                            <td class="code">
                                <div class="highlight">
                                    <pre class="hljs php"><span class="hljs-comment">// app/config/config.php</span>
$container-&gt;loadFromExtension(<span class="hljs-string">'framework'</span>, <span class="hljs-keyword">array</span>( <span class="hljs-string">'secret'</span> =&gt; <span class="hljs-string">'%secret%'</span>, <span class="hljs-string">'router'</span> =&gt; <span class="hljs-keyword">array</span>( <span class="hljs-string">'resource'</span> =&gt; <span class="hljs-string">'%kernel.root_dir%/config/routing.php'</span>, ), <span class="hljs-comment">// ...</span>
)); <span class="hljs-comment">// ...</span></pre>
                                </div>
                            </td>
                        </tr>
                    </table>
                </div>
            </div>
        </li>
    </ul>
</div>

<h2 id="field-variables">Field Variables<a class="headerlink" href="#field-variables" title="Permalink to this headline">¶</a></h2>
<table class="docutils" border="1">
    <thead valign="bottom">
    <tr class="row-odd">
        <th class="head">Variable</th>
        <th class="head">Type</th>
        <th class="head">Usage</th>
    </tr>
    </thead>
    <tbody valign="top">
    <tr class="row-even">
        <td>widget</td>
        <td><code class="notranslate">mixed</code></td>
        <td>The value of the <code class="notranslate">widget</code> option</td>
    </tr>
    <tr class="row-odd">
        <td>type</td>
        <td><code class="notranslate">string</code></td>
        <td>Multiple lines of text here, to show
that off</td>
    </tr>
    </tbody>
</table>
</body>
</html>